modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataKeyIdObjectResolver.java

```diff
@@ -7,102 +7,193 @@
 package org.mule.runtime.module.extension.internal.metadata;
 
 import static java.lang.String.format;
-import static java.util.function.Function.identity;
 import static org.mule.metadata.java.utils.JavaTypeUtils.getType;
 import static org.mule.runtime.api.metadata.resolving.FailureCode.INVALID_METADATA_KEY;
-import static org.mule.runtime.api.metadata.resolving.FailureCode.UNKNOWN;
-import static org.mule.runtime.module.extension.internal.util.IntrospectionUtils.checkInstantiable;
 import static org.mule.runtime.module.extension.internal.util.IntrospectionUtils.getAnnotatedFields;
+import static org.mule.runtime.module.extension.internal.util.IntrospectionUtils.getMetadataKeyParts;
+
 import org.mule.metadata.api.model.MetadataType;
+import org.mule.metadata.api.model.ObjectType;
+import org.mule.metadata.api.model.StringType;
+import org.mule.metadata.api.visitor.MetadataTypeVisitor;
 import org.mule.runtime.api.metadata.MetadataKey;
 import org.mule.runtime.api.metadata.MetadataResolvingException;
-import org.mule.runtime.core.util.collection.ImmutableMapCollector;
+import org.mule.runtime.core.api.component.Component;
+import org.mule.runtime.core.util.ValueHolder;
 import org.mule.runtime.extension.api.annotation.metadata.MetadataKeyId;
 import org.mule.runtime.extension.api.annotation.metadata.MetadataKeyPart;
 import org.mule.runtime.extension.api.introspection.ComponentModel;
+import org.mule.runtime.extension.api.introspection.metadata.NullMetadataKey;
+import org.mule.runtime.extension.api.introspection.parameter.ParameterModel;
 import org.mule.runtime.extension.api.introspection.property.MetadataKeyIdModelProperty;
 import org.mule.runtime.module.extension.internal.util.FieldSetter;
 
 import java.lang.reflect.Field;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
 
 /**
- * Provides an instance of the annotated {@link MetadataKeyId} parameter's complex type.
- * The instance will be populated with all the corresponding values of the passed {@link MetadataKey}.
+ * Provides an instance of the annotated {@link MetadataKeyId} parameter type.
+ * The instance will be populated with all the corresponding values of the passed {@link Map} key.
  *
  * @since 4.0
  */
 final class MetadataKeyIdObjectResolver
```

Comment:
don't use this class through static methods. Instantiate it and use it through regular methods. This is important to make it testable.